src/core/etl/src/Flow/ETL/PHP/Type/ArrayContentDetector.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Flow\ETL\PHP\Type;
+
+use Flow\ETL\Exception\InvalidArgumentException;
+use Flow\ETL\PHP\Type\Native\ArrayType;
+use Flow\ETL\PHP\Type\Native\NullType;
+use Flow\ETL\PHP\Type\Native\ScalarType;
+
+final class ArrayContentDetector
+{
+    private readonly ?Type $firstKeyType;
+
+    private readonly ?Type $firstValueType;
+
+    public function __construct(private readonly Types $uniqueKeysType, private readonly Types $uniqueValuesType)
+    {
+        $this->firstKeyType = $this->uniqueKeysType->first();
+        $this->firstValueType = $this->uniqueValuesType->first();
+    }
+
+    public function firstKeyType() : ?ScalarType
+    {
+        if (null !== $this->firstKeyType && !$this->firstKeyType instanceof ScalarType) {
+            throw InvalidArgumentException::because('First unique key type must be of ScalarType, given: ' . $this->firstKeyType::class);
+        }
+
+        return $this->firstKeyType;
+    }
+
+    public function firstValueType() : ?Type
+    {
+        return $this->firstValueType;
+    }
+
+    public function isList() : bool
+    {
+        if (!$this->firstKeyType()?->isInteger()) {
+            return false;
+        }
+
+        return 1 === $this->uniqueValuesType->without(new ArrayType(true), new NullType())->count();
```

Comment:
It's a small change but I would replace `new ArrayType(true)` with `ArrayType::empty()` to make it more descriptive